rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp

```diff
@@ -77,7 +77,7 @@ MultiThreadedExecutor::run(size_t)
       if (!rclcpp::ok(this->context_) || !spinning.load()) {
         return;
       }
-      if (!get_next_executable(any_exec)) {
+      if (!get_next_executable(any_exec, std::chrono::milliseconds(500))) {
```

Comment:
Maybe, we should refactor `MultiThreadedExecutor` to only one thread waiting and coordinating the work, and the others doing work.
@wjwwood any opinions?

Reply:
I didn't realize, but we are triggering a guard condition after executing any executable, which should awake the `wait_set`: https://github.com/ros2/rclcpp/blob/b76e5212b05e38000c4991aaf4dd86db893f7815/rclcpp/src/rclcpp/executor.cpp#L302-L306.
So, the timeout shouldn't be needed. @guillaumeautran Have you experienced a hang when using mutually exclusive callback groups?

Reply:
@ivanpauno Yes, I have experienced a hang. My situation was that I had two callback groups. The first (default) contained topics and timers and was mutually exclusive while the second callback group contained services only (and was reentrant). On a system with large enough number of CPU (10 core), the problem would occur where one thread would be processing the wait set and skipping the mutually exclusive callback group (as it was locked by another thread). The thread waiting on the incomplete waitset would never be woken up causing a infinite wait. 

I can most definitely see if there is a way to get the timeout user configurable.

Reply:
Ok, I get it.
There's a race condition, between one thread triggering the guard condition after execution and another thread waiting.
The guard condition may be triggered after the other thread has taken the `wait_mutex_` but it isn't waiting yet. In that case, it will hang forever.

I don't have an easy solution in mind (avoiding a timeout).

Reply:
We could potentially setup a protected boolean flag indicating the trigger has occurred. I do feel it is a bit of an overkill though but may be note. 

In the meantime, I've made the get_next_executable timeout configurable at instantiation of the multithreaded executor itself. Let me know what you think.

Reply:
> We could potentially setup a protected boolean flag indicating the trigger has occurred. I do feel it is a bit of an overkill though but maybe not.

I also thought about that, but I think that a race condition is still possible.
If you have a way to avoid the race condition with an extra boolean flag, I prefer that than a timeout.

I asked for someone else review, so you can wait to that.

I think that we should refactor the `MultiThreadedExecutor` to one thread waiting for work, and a thread pool doing work (`rclpy` is doing that). But that's another discussion.

Reply:
> In the meantime, I've made the get_next_executable timeout configurable at instantiation of the multithreaded executor itself. Let me know what you think.

The changes LGTM :+1: .

Reply:
Any update on the review for this? Code is starting to get stale....

Reply:
I also agree that adding a timeout isn't ideal. I think you would need a new lock to avoid the race condition you've described @ivanpauno. I think it should be possible to fix this race without any timeouts.

---

As for waiting in one thread and working in the others, I think the problem there is that you'd prefer not to wait at all when there are no threads available to do work. That's because you do not want to wait, see there is work, and then need to wait for a thread to do the work, because while waiting on a thread to become available the things you were going to work on may not need it any longer (message was discarded due to lifespan expiring or deadline being missed, or maybe a timer was canceled, etc.).

If we could do it so that you only ever wait once a thread is available (similar to how it happens now), then we could refactor it in this way.